openid/openid.go

```diff
@@ -0,0 +1,273 @@
+type Request struct {
+	// ReturnTo contains the callback address for the service, this is
+	// where the login response will come.
+	ReturnTo string
+
+	// Realm contains the realm that the user is logging into. See
+	// http://openid.net/specs/openid-authentication-2_0.html#realms
+	// for details.
+	Realm string
+
+	// Teams contains a list of launchpad teams to query membership
+	// of for the logged in user.
+	Teams []string
+
+	// SRegRequired contains a list of simple registration fields
+	// that are required by the service.
+	SRegRequired []string
+
+	// SRegOptional contains a list of simple registration fields
+	// that are optional, but requested by the service.
+	SRegOptional []string
+}
+
+// RedirectURL creates an OpenID login request addressed to c.Server.
+func (c Client) RedirectURL(r *Request) string {
+	v := url.Values{
+		"openid.ns":         {"http://specs.openid.net/auth/2.0"},
+		"openid.mode":       {"checkid_setup"},
+		"openid.claimed_id": {"http://specs.openid.net/auth/2.0/identifier_select"},
+		"openid.identity":   {"http://specs.openid.net/auth/2.0/identifier_select"},
+		"openid.return_to":  {r.ReturnTo},
+	}
+	if r.Realm != "" {
+		v.Set("openid.realm", r.Realm)
+	}
+	if len(r.Teams) > 0 {
+		v.Set("openid.ns.lp", nsTeams)
+		v.Set("openid.lp.query_membership", strings.Join(r.Teams, ","))
+	}
+	if len(r.SRegRequired) > 0 {
+		v.Set("openid.ns.sreg", nsSReg)
+		v.Set("openid.sreg.required", strings.Join(r.SRegRequired, ","))
+	}
+	if len(r.SRegOptional) > 0 {
+		v.Set("openid.ns.sreg", nsSReg)
+		v.Set("openid.sreg.optional", strings.Join(r.SRegOptional, ","))
+	}
+	if (c.Server == usso.UbuntuSSOServer{}) {
+		c.Server = usso.ProductionUbuntuSSOServer
+	}
+	return c.Server.OpenIDURL() + "?" + v.Encode()
+}
+
+// Response contains the values returned from Ubuntu SSO after a
+// successful login.
+type Response struct {
+	// ID contians the claimed_id of the logged in user. This will
+	// always be present in a successful login.
+	ID string
+
+	// Teams contains any launchpad teams that were specified in the
+	// OpenID response.
+	Teams []string
+
+	// SReg contains any simple registration fileds are
+	// were provided in the OpenID response.
+	SReg map[string]string
+}
+
+// verify is used to perform the OpenID verification of the login
+// response. This is declared as a variable so it can be overridden for
+// testing.
+var verify = openid.Verify
+
+// Verify processes a positive assertion from Ubuntu SSO. If the
+// verification is successful any parameters asserted by Ubuntu SSO will
+// be set in the Response. If the OpenID response reports that the login
+// was cancelled then an error will be returned with a cause of
+// ErrCancel. If the OpenID response reports an error occured then an
+// err will be returned with a cause fo ErrOpenID.
+func (c Client) Verify(requestURL string) (*Response, error) {
+	if (c.Server == usso.UbuntuSSOServer{}) {
```

Comment:
Rather than doing this here, how about having a NewClient function
instead that fills them in? Then perhaps you won't feel the need for
defaultNonceStore at all. It also potentially makes it easier for
people to layer their own stuff (logging, for example) on top of the
default values.